docs/conceptual/core/executor.md

```diff
@@ -2,8 +2,91 @@
+
+### Runtime Versioning
+
+In order for the executor to be able to select the appropriate runtime execution environment, it needs to know the name and version of both the native and Wasm runtime.
+
+The runtime provides the following versioning properties:
+
+- `spec_name`: The identifier for the different Substrate runtimes.
+
+- `impl_name`: The name of the implementation of the spec. This is of little consequence for the node and serves only to differentiate code of different implementation teams.
+
+- `authoring_version`: The version of the authorship interface. An authoring node will not attempt to author blocks unless this is equal to its native runtime.
+
+- `spec_version`: The version of the runtime specification. A full-node will not attempt to use its native runtime in substitute for the on-chain Wasm runtime unless all of `spec_name`, `spec_version`, and `authoring_version` are the same between Wasm and native.
+
+- `impl_version`: The version of the implementation of the specification. Nodes are free to ignore this; it serves only as an indication that the code is different; as long as the other two versions are the same then while the actual code may be different, it is nonetheless required to do the same thing. Non-consensus-breaking optimizations are about the only changes that could be made which would result in only the `impl_version` changing.
+
+As mentioned above, the executor always verifies that the native runtime has the same consensus-driven logic before it chooses to execute it, independent of whether the version is higher or lower.
+
+> **Note:** The runtime versioning is manually set. Thus the executor can still make inappropriate decisions if the runtime version is misrepresented.
+
+### Forkless Runtime Upgrades
+
+Traditional blockchains require a [hard fork](https://en.wikipedia.org/wiki/Fork_(blockchain)) when upgrading the state transition function of their chain. This requires node operators to stop their nodes and manually upgrade to the latest executable. For distributed production networks, coordination of a hard fork upgrades can be a complex process.
+
+The culmination of the properties listed on this page allows for Substrate-based blockchains to perform "forkless runtime upgrades". This means that the upgrade of the runtime logic can happen in real time without causing a fork in the network.
+
+To perform a forkless runtime upgrade, Substrate uses existing runtime logic to update the Wasm runtime stored on the blockchain to a new version with new logic. This upgrade gets pushed out to all syncing nodes on the network as a part of the consensus process. Once the Wasm runtime is upgraded, the executor will see that the native runtime version no longer matches this new Wasm runtime. As a result, it will fall back to execute the canonical Wasm runtime instead of using the native runtime in any of the execution processes.
```

Comment:
If you just update the `impl_version`, you can still use the native version, even after an upgrade.

Reply:
I clarified to say:

```
Once the Wasm runtime is upgraded, the executor will see that the native runtime `spec_name`, `spec_version`, or `authoring_version` no longer matches this new Wasm runtime.
```

The alternative was to say "consensus-breaking logic", but this is probably less explicit.